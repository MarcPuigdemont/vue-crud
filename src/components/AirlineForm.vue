<template>
  <b-form @submit="$event.preventDefault()" @reset="$event.preventDefault()" class="mx-2 airline-form">
    <b-form-row>
      <b-form-group id="input-group-1" label="IATA:" label-for="input-1" class="w-50 pr-1">
        <b-form-input id="input-1" v-model="form.iata" required placeholder="Enter IATA code" data-test="form-iata-input" />
        <b-form-invalid-feedback :state="form.iata.length > 0">
          IATA must be filled
        </b-form-invalid-feedback>
      </b-form-group>
      <b-form-group id="input-group-2" label="Airline Name:" label-for="input-2" class="w-50 pl-1">
        <b-form-input id="input-2" v-model="form.name" required placeholder="Enter airline name" data-test="form-name-input" />
        <b-form-invalid-feedback :state="form.name.length > 0">
          Name must be filled
        </b-form-invalid-feedback>
      </b-form-group>
    </b-form-row>
    <b-form-row>
      <b-form-group id="input-group-3" label="Primary Color:" label-for="input-3" label-class="crud-form-label" class="w-25 pr-1">
        <b-form-input id="input-3" v-model="form.primary_color" type="color" required placeholder="Enter primary color"></b-form-input>
      </b-form-group>

      <b-form-group id="input-group-4" label="Secondary Color:" label-for="input-4" label-class="crud-form-label" class="w-25 px-1">
        <b-form-input id="input-4" v-model="form.secondary_color" type="color" required placeholder="Enter secondary color"></b-form-input>
      </b-form-group>

      <b-form-group id="input-group-5" label="Services:" label-for="checkboxes-1" class="align-content-end w-50">
        <b-form-checkbox-group v-model="form.services" id="checkboxes-1" class="ml-2">
          <b-form-checkbox value="bags">
            <i class="material-icons" title="Bags" :disabled="isServiceDisabled('bags')" data-test="form-bags-checkbox">
              {{ icons['BAGS'] }}
            </i>
          </b-form-checkbox>
          <b-form-checkbox value="checkin">
            <i class="material-icons" title="Check-in" :disabled="isServiceDisabled('checkin')" data-test="form-checkin-checkbox">
              {{ icons['CHECKIN'] }}
            </i>
          </b-form-checkbox>
          <b-form-checkbox value="seats">
            <i class="material-icons" title="Seats" :disabled="isServiceDisabled('seats')" data-test="form-seats-checkbox">
              {{ icons['SEATS'] }}
            </i>
          </b-form-checkbox>
        </b-form-checkbox-group>
      </b-form-group>
    </b-form-row>
  </b-form>
</template>
<script>
import { BForm, BFormCheckbox, BFormCheckboxGroup, BFormGroup, BFormInput, BFormInvalidFeedback, BFormRow } from 'bootstrap-vue';
import constants from '@/constants';
import formatColor from '@/services/colorService.jsx';

export default {
  name: 'AirlineForm',
  components: {
    BForm,
    BFormCheckbox,
    BFormCheckboxGroup,
    BFormGroup,
    BFormInput,
    BFormInvalidFeedback,
    BFormRow,
  },
  props: {
    action: {
      type: String,
      required: true,
    },
    entity: Object,
  },
  data() {
    return {
      form: {
        iata: '',
        name: '',
        primary_color: '#293041',
        secondary_color: '#F138A4',
        services: [],
      },
      icons: constants.ICONS,
    };
  },
  computed: {
    isValid() {
      return !!(this.form.iata && this.form.name);
    },
  },
  mounted() {
    if (this.entity) {
      const airline = this.entity;
      const form = { ...airline };

      form.primary_color = formatColor(form.primary_color);
      form.secondary_color = formatColor(form.secondary_color);
      form.services = Object.keys(airline.services).filter(k => airline.services[k]);

      this.form = form;
    }
  },
  methods: {
    enitityUpdated() {
      const services = {
        bags: false,
        checkin: false,
        seats: false,
      };
      this.form.services.forEach(s => (services[s] = true));

      // Before spread operator: Object.assign({}, this.form, { services })
      return { ...this.form, ...{ services } };
    },
    isServiceDisabled(name) {
      return !this.form.services.find(s => s === name);
    },
  },
  watch: {
    form: {
      handler: function() {
        this.$emit('entityUpdated', this.enitityUpdated());
      },
      deep: true,
    },
    isValid: {
      handler: function() {
        this.$emit('validate', this.isValid);
      },
      immediate: true,
    },
  },
};
</script>
<style>
.crud-form-label {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}
</style>
