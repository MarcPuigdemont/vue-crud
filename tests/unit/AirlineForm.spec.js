import { mount } from '@vue/test-utils';
import AirlineForm from '@/components/AirlineForm.vue';

describe('AirlineForm', () => {
  test('should match the snapshot on create', () => {
    const propsData = {
      action: 'create',
    };
    const wrapper = mount(AirlineForm, {
      propsData,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  test('should match the snapshot on update', () => {
    const propsData = {
      action: 'upadte',
      entity: {
        iata: 'BX',
        name: 'Airline name',
        primary_color: '#ffffff',
        secondary_color: '#000000',
        services: { bags: true, checkin: false, seats: true },
      },
    };
    const wrapper = mount(AirlineForm, {
      propsData,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
