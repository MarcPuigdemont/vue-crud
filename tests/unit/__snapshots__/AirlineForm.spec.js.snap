// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirlineForm should match the snapshot on create 1`] = `
<form
  class="mx-2 airline-form"
>
  <div
    class="form-row"
  >
    <div
      class="form-group w-50 pr-1"
      id="input-group-1"
      role="group"
    >
      <label
        class="d-block"
        for="input-1"
        id="input-group-1__BV_label_"
      >
        IATA:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          data-test="form-iata-input"
          id="input-1"
          placeholder="Enter IATA code"
          required="required"
          type="text"
        />
         
        <div
          class="invalid-feedback d-block"
        >
          
        IATA must be filled
      
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      class="form-group w-50 pl-1"
      id="input-group-2"
      role="group"
    >
      <label
        class="d-block"
        for="input-2"
        id="input-group-2__BV_label_"
      >
        Airline Name:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          data-test="form-name-input"
          id="input-2"
          placeholder="Enter airline name"
          required="required"
          type="text"
        />
         
        <div
          class="invalid-feedback d-block"
        >
          
        Name must be filled
      
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
   
  <div
    class="form-row"
  >
    <div
      class="form-group w-25 pr-1"
      id="input-group-3"
      role="group"
    >
      <label
        class="d-block crud-form-label"
        for="input-3"
        id="input-group-3__BV_label_"
      >
        Primary Color:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          id="input-3"
          placeholder="Enter primary color"
          required="required"
          type="color"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      class="form-group w-25 px-1"
      id="input-group-4"
      role="group"
    >
      <label
        class="d-block crud-form-label"
        for="input-4"
        id="input-group-4__BV_label_"
      >
        Secondary Color:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          id="input-4"
          placeholder="Enter secondary color"
          required="required"
          type="color"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      class="form-group align-content-end w-50"
      id="input-group-5"
      role="group"
    >
      <label
        class="d-block"
        for="checkboxes-1"
        id="input-group-5__BV_label_"
      >
        Services:
      </label>
      <div>
        <div
          class="ml-2"
          id="checkboxes-1"
          role="group"
          tabindex="-1"
        >
          <div
            class="custom-control custom-control-inline custom-checkbox"
          >
            <input
              autocomplete="off"
              class="custom-control-input"
              name="checkboxes-1"
              type="checkbox"
              value="bags"
            />
            <label
              class="custom-control-label"
            >
              <i
                class="material-icons"
                data-test="form-bags-checkbox"
                disabled="disabled"
                title="Bags"
              >
                
            work
          
              </i>
            </label>
          </div>
           
          <div
            class="custom-control custom-control-inline custom-checkbox"
          >
            <input
              autocomplete="off"
              class="custom-control-input"
              name="checkboxes-1"
              type="checkbox"
              value="checkin"
            />
            <label
              class="custom-control-label"
            >
              <i
                class="material-icons"
                data-test="form-checkin-checkbox"
                disabled="disabled"
                title="Check-in"
              >
                
            where_to_vote
          
              </i>
            </label>
          </div>
           
          <div
            class="custom-control custom-control-inline custom-checkbox"
          >
            <input
              autocomplete="off"
              class="custom-control-input"
              name="checkboxes-1"
              type="checkbox"
              value="seats"
            />
            <label
              class="custom-control-label"
            >
              <i
                class="material-icons"
                data-test="form-seats-checkbox"
                disabled="disabled"
                title="Seats"
              >
                
            airline_seat_recline_normal
          
              </i>
            </label>
          </div>
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</form>
`;

exports[`AirlineForm should match the snapshot on update 1`] = `
<form
  class="mx-2 airline-form"
>
  <div
    class="form-row"
  >
    <div
      class="form-group w-50 pr-1"
      id="input-group-1"
      role="group"
    >
      <label
        class="d-block"
        for="input-1"
        id="input-group-1__BV_label_"
      >
        IATA:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          data-test="form-iata-input"
          id="input-1"
          placeholder="Enter IATA code"
          required="required"
          type="text"
        />
         
        <div
          class="invalid-feedback"
        >
          
        IATA must be filled
      
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      class="form-group w-50 pl-1"
      id="input-group-2"
      role="group"
    >
      <label
        class="d-block"
        for="input-2"
        id="input-group-2__BV_label_"
      >
        Airline Name:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          data-test="form-name-input"
          id="input-2"
          placeholder="Enter airline name"
          required="required"
          type="text"
        />
         
        <div
          class="invalid-feedback"
        >
          
        Name must be filled
      
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
   
  <div
    class="form-row"
  >
    <div
      class="form-group w-25 pr-1"
      id="input-group-3"
      role="group"
    >
      <label
        class="d-block crud-form-label"
        for="input-3"
        id="input-group-3__BV_label_"
      >
        Primary Color:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          id="input-3"
          placeholder="Enter primary color"
          required="required"
          type="color"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      class="form-group w-25 px-1"
      id="input-group-4"
      role="group"
    >
      <label
        class="d-block crud-form-label"
        for="input-4"
        id="input-group-4__BV_label_"
      >
        Secondary Color:
      </label>
      <div>
        <input
          aria-required="true"
          class="form-control"
          id="input-4"
          placeholder="Enter secondary color"
          required="required"
          type="color"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      class="form-group align-content-end w-50"
      id="input-group-5"
      role="group"
    >
      <label
        class="d-block"
        for="checkboxes-1"
        id="input-group-5__BV_label_"
      >
        Services:
      </label>
      <div>
        <div
          class="ml-2"
          id="checkboxes-1"
          role="group"
          tabindex="-1"
        >
          <div
            class="custom-control custom-control-inline custom-checkbox"
          >
            <input
              autocomplete="off"
              class="custom-control-input"
              name="checkboxes-1"
              type="checkbox"
              value="bags"
            />
            <label
              class="custom-control-label"
            >
              <i
                class="material-icons"
                data-test="form-bags-checkbox"
                title="Bags"
              >
                
            work
          
              </i>
            </label>
          </div>
           
          <div
            class="custom-control custom-control-inline custom-checkbox"
          >
            <input
              autocomplete="off"
              class="custom-control-input"
              name="checkboxes-1"
              type="checkbox"
              value="checkin"
            />
            <label
              class="custom-control-label"
            >
              <i
                class="material-icons"
                data-test="form-checkin-checkbox"
                disabled="disabled"
                title="Check-in"
              >
                
            where_to_vote
          
              </i>
            </label>
          </div>
           
          <div
            class="custom-control custom-control-inline custom-checkbox"
          >
            <input
              autocomplete="off"
              class="custom-control-input"
              name="checkboxes-1"
              type="checkbox"
              value="seats"
            />
            <label
              class="custom-control-label"
            >
              <i
                class="material-icons"
                data-test="form-seats-checkbox"
                title="Seats"
              >
                
            airline_seat_recline_normal
          
              </i>
            </label>
          </div>
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</form>
`;
